lcls-twincat-motion/Library/POUs/Motion/States/FB_StateSetupHelper.TcPOU

```diff
@@ -0,0 +1,107 @@
+    from the input stPositionState. Note that the other args will be ignored in this case.
+
+    Calling without bSetDefault or with it set to FALSE will apply values to the
+    input stPositionState with the following priority order:
+
+    1. The value used in the function block call
+    2. The value from the template stPositionState used in the most recent
+       call with bSetDefault:=TRUE
+    3. The original default value as defined on ST_PositionState
+
+    Example expected usage:
+    VAR
+        fbStateSetup: FB_StateSetupHelper;
+        stDefault: ST_PositionState := (
+            fDelta := 0.5,
+            fVelocity := 10,
+            bMoveOk := TRUE,
+            bValid := TRUE
+        );
+        astStates1: ARRAY[1..GeneralConstants.MAX_STATES] OF ST_PositionState;
+        astStates2: ARRAY[1..GeneralConstants.MAX_STATES] OF ST_PositionState;
+        astStates2: ARRAY[1..GeneralConstants.MAX_STATES] OF ST_PositionState;
+    END_VAR
+
+    fbStateSetup(stPositionState:=stDefault, bSetDefault:=TRUE);
+
+    fbStateSetup(stPositionState:=astStates1[1], sName:='OUT', fPosition:=10);
+    fbStateSetup(stPositionState:=astStates1[2], sName:='YAG', fPosition:=20);
+    fbStateSetup(stPositionState:=astStates1[3], sName:='TT', fPosition:=30);
+
+    fbStateSetup(stPositionState:=astStates2[1], sName:='OUT', fPosition:=-30);
+    fbStateSetup(stPositionState:=astStates2[2], sName:='YAG', fPosition:=35);
+    fbStateSetup(stPositionState:=astStates2[3], sName:='TT', fPosition:=70);
+
+    fbStateSetup(stPositionState:=astStates3[1], sName:='OUT', fPosition:=0.4, fDelta:=0.1);
+    fbStateSetup(stPositionState:=astStates3[2], sName:='YAG', fPosition:=2.3, fDelta:=0.1);
+    fbStateSetup(stPositionState:=astStates3[3], sName:='TT', fPosition:=5.6, fDelta:=0.1;
+*)
+VAR_IN_OUT
+    stPositionState: ST_PositionState;
+END_VAR
+VAR_INPUT
+    bSetDefault: BOOL;
+    sName: STRING;
+    fPosition: LREAL;
+    nEncoderCount: UDINT;
+    fDelta: LREAL;
+    fVelocity: LREAL;
+    fAccel: LREAL;
+    fDecel: LREAL;
+    bMoveOk: BOOL;
+    bLocked: BOOL;
+    bValid: BOOL;
+    bUseRawCounts: BOOL;
+    sPmpsState: STRING;
+END_VAR
+VAR
+    stDefault: ST_PositionState;
+END_VAR
+]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[
+IF bSetDefault THEN
+    bSetDefault := FALSE;
+    stDefault := stPositionState;
+ELSE
+    stPositionState.sName := sName;
+    stPositionState.fPosition := fPosition;
+    stPositionState.nEncoderCount := nEncoderCount;
+    stPositionState.fDelta := fDelta;
+    stPositionState.fVelocity := fVelocity;
+    stPositionState.fAccel := fAccel;
+    stPositionState.fDecel := fDecel;
+    stPositionState.bMoveOk := bMoveOk;
+    stPositionState.bLocked := bLocked;
+    stPositionState.bValid := bValid;
+    stPositionState.bUseRawCounts := bUseRawCounts;
+    stPositionState.stPMPS.sPmpsState := sPmpsState;
+END_IF
+
+// Overwrite the input args so that unset args are the defaults in the next call
```

Comment:
If the user forgets or decides not call `bSetDefault` this will use the "default defaults" as defined by the `ST_PositionState` struct, which are mostly `0`/`FALSE`/`''` (except the default name is `'Invalid'`).

Before doing something, I need to ask a quick question: are there any situations where we don't want to set a default? I think the answer is "no", you always want to at least opt in to `bValid:=TRUE`. So I agree that we should do something here.

I think the most prudent response is to make a warning message? Should I also force `bValid:=FALSE` so the state can't be used in the normal ways?

Reply:
My first thought was a warning message like you said, but they are somewhat heavy/annoying to use (in my opinion) for things which might get called repeatedly like this.

`No bSetDefault call -> bValid:=FALSE` seems like it might be enough in this scenario, since it won't allow motion and should cover the majority of mistaken copy/paste errors.

Reply:
![image](https://github.com/pcdshub/lcls-twincat-motion/assets/10647860/58cd8cc8-9e00-40bd-8d9c-c4989b115bd1)

Reply:
Well that looks awesome! Nicely done (with the `bHasWarned` stuff too)

Reply:
I tried to resolve this with https://github.com/pcdshub/lcls-twincat-motion/pull/184/commits/2d55fa2d20a6f75e5579276fa11a031197883ac4

Reply:
I considered for a moment if I can include other info in the payload, but all I have custody of is the state name, I don't have access to the motor name for example. I'm not sure if it's worth including that info in the warning message.

Reply:
Revising my brain a bit: I can include the pmps string (if provided) which would help someone track down the FB.